corporate/lib/stripe.py

```diff
@@ -1096,25 +1121,58 @@ def process_initial_upgrade(
                             _("Please add a credit card before starting your free trial."),
                         )
 
+            event_time = billing_cycle_anchor
+            if should_schedule_upgrade_for_legacy_remote_server:  # nocoverage
+                assert remote_server_legacy_plan is not None
+                if charge_automatically:
+                    remote_server_legacy_plan.charge_automatically = True
+                    # Ensure customers not paying via invoice have a default payment method set.
+                    stripe_customer = stripe_get_customer(customer.stripe_customer_id)
+                    if not stripe_customer_has_credit_card_as_default_payment_method(
+                        stripe_customer
+                    ):
+                        raise BillingError(
+                            "no payment method",
+                            _("Please add a credit card to schedule upgrade."),
+                        )
+
+                # Settings status > CustomerPLan.LIVE_STATUS_THRESHOLD makes sure we don't have
+                # to worry about this plan being used for any other purpose.
+                # NOTE: This is the 2nd plan for the customer.
+                plan_params["status"] = CustomerPlan.NEVER_STARTED
+                event_time = timezone_now().replace(microsecond=0)
+
+                # Switch to the new plan at plan end date.
+                assert remote_server_legacy_plan.end_date == billing_cycle_anchor
+                remote_server_legacy_plan.status = CustomerPlan.SWITCH_PLAN_TIER_AT_PLAN_END
+                remote_server_legacy_plan.price_per_license = price_per_license
+                remote_server_legacy_plan.save(
+                    update_fields=["status", "charge_automatically", "price_per_license"]
+                )
+            elif remote_server_legacy_plan is not None:  # nocoverage
+                remote_server_legacy_plan.status = CustomerPlan.ENDED
+                remote_server_legacy_plan.save(update_fields=["status"])
+
             plan = CustomerPlan.objects.create(
                 customer=customer, next_invoice_date=next_invoice_date, **plan_params
             )
             ledger_entry = LicenseLedger.objects.create(
                 plan=plan,
                 is_renewal=True,
-                event_time=billing_cycle_anchor,
+                event_time=event_time,
                 licenses=billed_licenses,
                 licenses_at_next_renewal=billed_licenses,
             )
```

Comment:
Why do we create license ledger entries against the newly created NEVER_STARTED plan in the `should_schedule_upgrade_for_legacy_remote_server` case? We create license ledger entries inside `make_end_of_cycle_updates_if_needed`, and those seem more correct as they will use the data available at the right date...

Reply:
I added a couple comments noting that details like this feels like a hack; I don't quite understand the implications of this and it's probably worth discussing in chat regarding follow-ups for this PR -- I'm kinda thinking something like first-class support for sending data on all non-ended plans to the billing page template so that it can display both plans in a more natural way is probably the right path forward here.

But I think we can merge this hack just to let us keep building other pieces on top of this work.